### ISSUE TITLE
Failed build notifications not firing after shell script failure

### ISSUE THREAD

**Issue by artgon** [jenkinsci/slack-plugin#134]
We have a shell script that runs our regression tests. If the tests fail, the shell script exits with `exit -1`. When this happens the failed build alert does not trigger. 

Here is a log sample (tried to sanitize as legibly as possible):

```
Started by user ****** (<https://****.com/job/****/1157/|Open>) danger
Sep 23, 2015 9:22:34 PM INFO jenkins.plugins.slack.StandardSlackService publish
Posting succeeded
Sep 23, 2015 9:22:34 PM INFO jenkins.plugins.slack.ActiveNotifier getChanges
Empty change...
Sep 23, 2015 9:22:34 PM INFO jenkins.plugins.slack.StandardSlackService publish
Posting: to #***** on ***** using https://*****.slack.com/services/hooks/jenkins-ci?token=***********: ******** - #1157 Starting... after 6.8 sec and counting (<https://******/job/*****/1157/|Open>) danger
Sep 23, 2015 9:22:34 PM INFO jenkins.plugins.slack.StandardSlackService publish
Posting succeeded
```

Only the starting messages went out, but not the failed messages.

**Comment by samrocketman** [comment]
Hello, I can't replicate the behavior.  I created a basic job with slack configured with a shell script build step containing the following code.

```bash
exit -1
```

Slack was successfully notified with a failed build.  Test environment includes:

* Mac OS X 10.9.5
* Jenkins ver. 1.630
* Slack plugin ver. 1.8

**Comment by artgon** [comment]
Yeah you're right. I think I've narrowed it down changing output color and then exiting. I've been able to reproduce it on a brand new job with this script:

```
echo
echo -e "\033[31mError!"
exit -1
```

Let me know if you can reproduce.

**Comment by samrocketman** [comment]
Confirmed, I can reproduce with that error.  When I configure a job with:

```bash
echo
echo -e "\033[31mError!"
exit -1
```
as a shell script it will show a failed status in Jenkins but not post to slack.  Very odd indeed.  Thanks for your work.  I'll mark this as a verified bug.

**Comment by mausch** [comment]
I'm apparently seeing this bug as well... all my build steps are shell scripts and I don't get any notifications about failures but the test connection works fine.

Jenkins does seem to recognise failures as such, i.e. in the build log I see:

```
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

However I don't have any colours in my output...

If I wanted to fix this would I have to start by debugging https://github.com/jenkinsci/slack-plugin/blob/master/src/main/java/jenkins/plugins/slack/ActiveNotifier.java#L87 ? Is that a reasonable place to start?

Thanks

**Comment by samrocketman** [comment]
@mausch you'll have to be more detailed about how your job is configured in order to help tracking down where to start.  If I can't reproduce your conditions then there's no real good answer with where to start other than digging into the source with how your job is configured.

**Comment by seregasheypak** [comment]
Hi, is there any workaround here?

**Comment by loyalty-company** [comment]
Same here, when test build failed I got no notification, wich is sad because it's there I want to be alerted (and less when it succed ^^). 
Any news ?

**Comment by seregasheypak** [comment]
BTW, it happens when we use slack-plugin after job with script. Plugin works correctly after mvn clean install for example :)

**Comment by kuboj** [comment]
Hello. I initially thought that I'm also affected by this bug (Slack _sometimes_ doesn't report failure of failed Ansible playbook execution, which has, by default, colored output) but I'm not able to reproduce it using clean job and your bash script
```bash
echo
echo -e "\033[31mError!"
exit -1
```
@artgon, @samrocketman Which Slack notifications have you enabled? Mine are:
```
Notify Failure		
Notify Success		
Notify Unstable		
Notify Back To Normal		
Notify Repeated Failure	
```

Versions:
Jenkins 1.644
slack-plugin 1.8.1

**Comment by bluepioupiou** [comment]
hey all, nothing new ?

**Comment by kennr** [comment]
I am also seeing this issue with the following configuration:

- Jenkins 1.625.2
- Mac OS X 10.10.5
- Java JDK 1.8.0_74
- Slack-plugin 1.8.1

If I have the following enabled...
```
Notify Build Now
Notify Failure
Notify Success
```
I get the following results in Slack:
> Job Name - #1 Started by user ken (Open)
> Job Name - #1 Starting... after 17 sec and counting (Open)

(this last one is the "failure")

However, if I have the following enabled...
```
Notify Failure
Notify Success
```
I only get a notification for Success, but nothing for Failure.

**Comment by kennr** [comment]
Found the culprit and it looks like it's intentional. In [this file](https://github.com/kennr/slack-plugin/blob/master/src/main/java/jenkins/plugins/slack/ActiveNotifier.java), if I comment out line 100 `&& previousResult != Result.FAILURE` and rebuild the plugin, I get Slack notifications for Failed builds that had a prior failure. This is the logic that was in HipChat and that I'd expect. I'll submit a PR, but I'm not sure if it'll get accepted since the comment before it is "//notify only on single failed build".

**Comment by samrocketman** [comment]
@kennr I believe I wrote that portion.  There was another issue (which I'll have to dig around for) where someone reported the expected/wanted behavior was to notify only on the first failure but not to continue notify on failures.  If one wanted to notify failures after the first build then one would have to check "notify repeated failure".

So to get notified for all failures both notify on failure and notify on repeated failures must be checked.

**Comment by angelblade27** [comment]
I am facing the same issue with slack-plugin 1.8.0

**Comment by doloreszhang** [comment]
any update? i met this issue as well

**Comment by rpherrera** [comment]
updates please!

**Comment by win-t** [comment]
Hi, I got same issue,
the problem somehow is not consistent, I tried to run job twice (no change are made), one time the fail notification was sent but in another time it is not sent at all.
any update on this issue? or anyone else found better slack plugin for jenkins?

**Comment by win-t** [comment]
maybe related #134

**Comment by pedropapa** [comment]
I had the same problem.. I fixed that configuring slack notifications **again** per project. Seems like global slack notifications are not being used in projects.

**Comment by win-t** [comment]
after reading https://github.com/jenkinsci/slack-plugin/blob/9c57909027c04261417d984dd87e8184012062ac/src/main/java/jenkins/plugins/slack/ActiveNotifier.java#L116,
I just know that Jenkins only notify for the first failure, and there is `NotifyRepeatedFailure` option in the advanced option to keep Jenkins notify repeated failure.
sorry I miss that one.

**Comment by glmnet** [comment]
What I'm doing here... can you rephrase the checkboxes?
Take out the "Notify Repeated Failure" from Hidden "Advanced" stuff, put them next to each other
Rename "Notify Failure" to "Notify Failure after success only"
Rename "Notify Repeated Failure" to "Notify every failure"

Thank you!

**Comment by deerawan** [comment]
After I checked box "Notify Repeated Failure", I got the slack notification. This solution works for me.

**Comment by timja** [comment]
I'm going to close this as working as intended, if someone wants to send a PR to tweak the label text feel free...
